### PR TITLE
Support for SQL.

### LINKED ISSUE #2683: spring-data-elasticsearch doesn't support elasticsearch sql
elasticsearch sql like this:

select count(1) from "indexPrefix*"

### PR DESCRIPTION
**Pull request by youssef3wi**
- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] **There is a ticket in the bug tracker for the project in our [issue tracker](https://github.com/spring-projects/spring-data-elasticsearch/issues)**. Add the issue number to the _Closes #issue-number_ line below
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

Closes #2683

### CODE REVIEW

**Comment by youssef3wi** [comment]
I should add some test case to verify complex results such as `SELECT SUM(views) FROM x`.

**Review by sothawo — CHANGES_REQUESTED** [review]
thanks a lot for this; just some minor remarks/questions

File: src/main/java/org/springframework/data/elasticsearch/client/elc/ElasticsearchTemplate.java [hunk 1]
```diff
@@ -16,13 +16,16 @@
 package org.springframework.data.elasticsearch.client.elc;
 
 import static org.springframework.data.elasticsearch.client.elc.TypeUtils.*;
+import static org.springframework.data.elasticsearch.core.sql.types.ResponseFormat.json;
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/client/elc/ElasticsearchTemplate.java:19]: this does not compile, that package does not exist anymore

File: src/main/java/org/springframework/data/elasticsearch/client/elc/ReactiveElasticsearchTemplate.java [hunk 2]
```diff
@@ -17,6 +17,7 @@
 
 import static co.elastic.clients.util.ApiTypeHelper.*;
 import static org.springframework.data.elasticsearch.client.elc.TypeUtils.*;
+import static org.springframework.data.elasticsearch.core.sql.types.ResponseFormat.json;
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/client/elc/ReactiveElasticsearchTemplate.java:20]: this does not compile, that package does not exist anymore

File: src/main/java/org/springframework/data/elasticsearch/client/elc/RequestConverter.java [hunk 3]
```diff
@@ -530,6 +531,22 @@ public co.elastic.clients.elasticsearch.indices.GetTemplateRequest indicesGetTem
...
+		Assert.notNull(query, "Query must not be null.");
+
+		return co.elastic.clients.elasticsearch.sql.QueryRequest.of(sqb -> {
+			sqb.query(query.getQuery()).catalog(query.getCatalog()).columnar(query.getColumnar()).cursor(query.getCursor())
+					.fetchSize(query.getFetchSize()).fieldMultiValueLeniency(query.getFieldMultiValueLeniency())
+					.indexUsingFrozen(query.getIndexIncludeFrozen()).keepAlive(time(query.getKeepAlive()))
+					.keepOnCompletion(query.getKeepOnCompletion()).pageTimeout(time(query.getPageTimeout()))
+					.requestTimeout(time(query.getRequestTimeout()))
+					.waitForCompletionTimeout(time(query.getWaitForCompletionTimeout())).filter(getQuery(query.getFilter(), null))
+					.timeZone(Objects.toString(query.getTimeZone(), null)).format(Objects.toString(query.getFormat(), null));
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/client/elc/RequestConverter.java:544]:
```suggestion
					.timeZone(Objects.toString(query.getTimeZone(), null)).format("json);
```

File: src/main/java/org/springframework/data/elasticsearch/client/elc/ElasticsearchTemplate.java [hunk 4]
```diff
@@ -656,6 +663,21 @@ public boolean deleteScript(String name) {
 		DeleteScriptRequest request = requestConverter.scriptDelete(name);
 		return execute(client -> client.deleteScript(request)).acknowledged();
 	}
+
+	@Override
+	public SqlResponse search(SqlQuery query) {
+		Assert.notNull(query, "Query must not be null.");
+		Assert.isTrue(query.getFormat() == null || json.equals(query.getFormat()),
+				"The Elasticsearch Java Client only supports JSON format.");
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/client/elc/ElasticsearchTemplate.java:671]:
```suggestion
```
nothing to assert here anymore

File: src/main/java/org/springframework/data/elasticsearch/client/elc/ReactiveElasticsearchTemplate.java [hunk 5]
```diff
@@ -646,6 +651,16 @@ public BaseQueryBuilder queryBuilderWithIds(List<String> ids) {
 		return NativeQuery.builder().withIds(ids);
 	}
 
+	@Override
+	public Mono<SqlResponse> search(SqlQuery query) {
+		Assert.notNull(query, "Query must not be null.");
+		Assert.isTrue(query.getFormat() == null || json.equals(query.getFormat()),
+				"The Elasticsearch Java Client only supports JSON format.");
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/client/elc/ReactiveElasticsearchTemplate.java:658]:
```suggestion
```

File: src/test/java/org/springframework/data/elasticsearch/core/sql/SqlOperationsIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.core.sql;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.springframework.data.elasticsearch.core.sql.types.ResponseFormat.yaml;
```
Comment (sothawo) [inline src/test/java/org/springframework/data/elasticsearch/core/sql/SqlOperationsIntegrationTests.java:23]:
```suggestion
```

File: src/test/java/org/springframework/data/elasticsearch/core/sql/SqlOperationsIntegrationTests.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		// Given
+		SqlQuery query = SqlQuery.builder("SELECT * FROM entity_for_sql WHERE views = 0").withFormat(yaml).build();
+
+		// When
+
+		// Then
+		assertThrows(IllegalArgumentException.class, () -> operations.search(query),
+				"The Elasticsearch Java Client only supports JSON format.");
+	}
+
```
Comment (sothawo) [inline src/test/java/org/springframework/data/elasticsearch/core/sql/SqlOperationsIntegrationTests.java:103]:
```suggestion
```

File: src/test/java/org/springframework/data/elasticsearch/core/sql/SqlOperationsIntegrationTests.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+		}
+	}
+
+	// begin configuration region
+	@Configuration
+	@Import({ ElasticsearchTemplateConfiguration.class })
+	static class Config {}
+	// end region
+
+	@Test
```
Comment (sothawo) [inline src/test/java/org/springframework/data/elasticsearch/core/sql/SqlOperationsIntegrationTests.java:78]:
```suggestion
	@Test // #2683
```

please add the issue number to the tests, overlooked that on the first review

**Review by sothawo — CHANGES_REQUESTED** [review]
doesn't compile 😁

**Pull request merged**